### ISSUE TITLE
UI not completely redrawn when returning from mame

### ISSUE THREAD

**Issue by JetA2** [zpaolo11x/Arcadeflow#51]
When returning from mame, only the thumbnail of the selected game is drawn. When navigating the UI, the thumbnails are redrawn as the selection square passes over them.

Using X on Arch Linux, AM+ and AF 17.4

It sounds a lot like this old issue, but it doesn't happen in other layouts, only in AF:
https://github.com/oomek/attractplus/issues/32

**Comment by zpaolo11x** [comment]
> When returning from mame, only the thumbnail of the selected game is drawn. When navigating the UI, the thumbnails are redrawn as the selection square passes over them.

You men the other thumbnails are blank? Do you have caching enabled in AM options? Can you share the console log when the issue happens?

**Comment by JetA2** [comment]
Image cache size is zero in AM settings.

I made some screenshots:

Immediately after returning from mame:
![sc1](https://github.com/user-attachments/assets/def79bbc-38e4-451d-8f63-07fbd8b09d43)

After pressing "right" once (when navigating further, more thumbnails appear):
![sc2](https://github.com/user-attachments/assets/ef216fc0-8906-49ba-b104-364bf3757ae0)

Here's the log, no errors that I can see:
```
Attract-Mode Plus v3.0.9 (Linux, SFML 2.6.0 +Xinerama +7z +Curl) 
avcodec 61.3.100 / avformat 61.1.100 / swscale 8.1.100 / avutil 59.8.100 / swresample 5.1.100

Config: /home/icade/.attract/attract.cfg

*** Initializing display: 'mame'
 - Loaded master romlist 'mame' in 42 ms (721 entries kept, 12997 discarded)
 - Constructed 44 filters in 5 ms (31724 comparisons)
 - Loaded layout: /home/icade/.attract/layouts/Arcadeflow-17.4/ (layout.nut)
 - Working directory: /home/icade/.mame/
*** Running: /usr/bin/mame vaportra
```

Apart from this, everything works great and it's an awesome frontend!

**Comment by zpaolo11x** [comment]
Ok, I cannot check on my rig at the moment, but could you try to go in AF options, Performance and FX, and enable the raspberry Pi fix? That should force a reload of the layout and bypass the issue, until I find a way to fix it

**Comment by JetA2** [comment]
The raspberry pi fix does bypass the issue.

**Comment by zpaolo11x** [comment]
Checked on my macBook and I see no issues, tomorrow I'll check with Ubuntu. Have you tried setting NonBLockingModeWait in AM options at a larger value?

**Comment by JetA2** [comment]
NonBlockingModeWait does make the layout reload when returning, but it has some side effects.
The first time mame is launched, it seems to be behind the frontend. All I see is the empty AF background but I can press escape to return to AF. AF is then in a window instead of full screen.
I can launch mame normally after that but AF remains in a window.

**Comment by zpaolo11x** [comment]
This is an Arch related issue, I just tried with Ubuntu 24.04 and although it has tons of problems (like audio in Retroarch not working, fullscreen not being consistent and stupid popups when retroarch starts) there is no issue when coming back from a game. Are you on Groovy Arcade?

**Comment by JetA2** [comment]
Ok, thanks for testing. I'm not using GroovyArcade, just a fresh Arch Linux install and the Arch mame package with AM+ compiled on the same machine.

I'm running sway as window manager (wayland). I can set mame to run with wayland, but attractplus still falls back to X11 using Xwayland. I'd like to test running everything with the wayland driver. Is there a way to specify the video driver with AM? 

Otherwise it might be a graphics hardware driver issue. The i915 driver is the one in use on my machine.

**Comment by JetA2** [comment]
I've been looking at the code and trying different things, and it seems like the multifilter code may have something to do with  it. When I comment out the multifilter code in `on_transition`, the issue goes away:

```
	if (ttype == Transition.FromGame) {
		if (prf.RPI) fe.set_display(fe.list.display_index)

		update_thumbdecor(focusindex.new, 0, tilez[focusindex.new].AR.current)
		/*
		mfz_build(true)
		try {
			mfz_load()
			mfz_populatereverse()
		} catch(err) {}
		mfz_apply(false)
		*/
	}
```

It's interesting, because I have another issue with multifilter:
When I bring up the multifilter menu and select a category, I don't get the next menu level. Instead the game selected "under" the menu starts right away. I'm guessing this issue doesn't show up on Ubuntu either?

**Comment by zpaolo11x** [comment]
> I've been looking at the code and trying different things, and it seems like the multifilter code may have something to do with it. When I comment out the multifilter code in `on_transition`, the issue goes away:

Thanks for spotting it, I'll have a look at the code. There is indeed an issue with Multifilter in AF 17.4 and that's why you see this:

> It's interesting, because I have another issue with multifilter: When I bring up the multifilter menu and select a category, I don't get the next menu level. Instead the game selected "under" the menu starts right away. I'm guessing this issue doesn't show up on Ubuntu either?

... this is a bug, I already fixed it for 17.5 but it's still not ready for release. You can check installing the 175_wip branch, you should not have the multifilter running the game issue, maybe it will help also in the "back to game" issue?

**Comment by JetA2** [comment]
Nice, 175_wip indeed fixes the multifilter issue.
The "back to game" issue still remains though.

Take your time, I really enjoy AF, very nice work!

**Comment by zpaolo11x** [comment]
Can you check if this fix works for you? Change the code you already found to be problematic with this one:

```
	if (ttype == Transition.FromGame) {
		if (prf.RPI) fe.set_display(fe.list.display_index)

		update_thumbdecor(focusindex.new, 0, tilez[focusindex.new].AR.current)
		mfz_build(true)
		try {
			mfz_load()
			mfz_populatereverse()
		} catch(err) {}
		mfz_apply(true)
	}
```

I'm investigating the source of the issue, but it seems with mfz_apply(true) it forces a tiles redesign and fixes the issue...

**Comment by JetA2** [comment]
Awesome, that change fixed the issue!

**Comment by JetA2** [comment]
I came across some information that may help with debugging.
I was running 17.4 with the multi filter code commented out like in https://github.com/zpaolo11x/Arcadeflow/issues/51#issuecomment-2466731150
Returning from mame worked fine in this config.

When I added a filter for "Most Played" to `attract.cfg` this brought back the "return from mame" issue.
After some testing, it seems that any filter containing `sort_by PlayedTime` triggers the issue, even with the multi filter code commented out.

**Comment by JetA2** [comment]
Setting AM to "Fill Screen" instead of "Fullscreen" seems to fix the issue without any changes to the layout.
I'm not sure what the difference is between the two settings.

**Comment by zpaolo11x** [comment]
I definitely have to check it out again because I was sure I tested with GroovyArcade both in Fill Screen and Fullscreen mode, but with the recent changes of fullscreen management in Ubuntu maybe something similar happened on Arch... As far as I know "fill screen" is just a window as large as the screen (actually 1 pixel larger in each direction) on top of everything.

When you talk about Most Played and sort by played time, are this AM filters? AF can be a bit picky on them, never used those myself so I'll have to do some further tests. I think it's plausible that AM filters might interfere with AF multifilter at this point. Will check asap...

**Comment by JetA2** [comment]
Yes, the Most Played filter is an AM filter defined in `attract.cfg`.
The filter works great in AF as well.
